File: PrefixSum/560.subarray-sum-equals-k.go
package leetcode

// @leet start
func subarraySum(nums []int, k int) int {
	var (
		preSum, res int
		preSCount   = map[int]int{0: 1}
	)
	for _, v := range nums {
		preSum += v
		// NOTE: diff := k - preSum // makesure the order
		diff := k - preSum
		if count, ok := preSCount[diff]; ok {
			res += count
		}
		preSCount[preSum]++
	}
	return res
}

// @leet end

File: PrefixSum/560.subarray-sum-equals-k.py
from typing import List


# @leet start
class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        total = res = 0
        preSum = {0: 1}
        for v in nums:
            total += v
            diff = total - k
            if diff in preSum:
                res += preSum[diff]
            preSum[total] = 1 + preSum.get(total, 0)
        return res


# @leet end

